include/rcutils/types/array_list.h

```diff
@@ -0,0 +1,208 @@
+ * rcutils_ret_t ret = rcutils_array_list_init(&array_list, 2, &allocator);
+ * if (ret != RCUTILS_RET_OK) {
+ *   // ... error handling
+ * }
+ * array_list.data[0] = rcutils_strdup("Hello", &allocator);
+ * array_list.data[1] = rcutils_strdup("World", &allocator);
+ * ret = rcutils_array_list_fini(&array_list);
+ *
+ * \param[inout] array_list object to be initialized
+ * \param[in] initial_capacity the initial capacity to allocate in the list
+ * \param[in] data_size the size (in bytes) of the data object being stored in the list
+ * \param[in] allocator to be used to allocate and deallocate memory
+ * \return `RCUTILS_RET_OK` if successful, or
+ * \return `RCUTILS_RET_INVALID_ARGUMENT` for invalid arguments, or
+ * \return `RCUTILS_RET_BAD_ALLOC` if memory allocation fails, or
+ * \return `RCUTILS_RET_ERROR` if an unknown error occurs
+ * ```
+ */
+RCUTILS_PUBLIC
+rcutils_ret_t
+rcutils_array_list_init(
+  rcutils_array_list_t * array_list,
+  size_t initial_capacity,
+  size_t data_size,
+  const rcutils_allocator_t * allocator);
+
+/// Finalize a string array, reclaiming all resources.
+/**
+ * This function reclaims any memory owned by the string array, including the
+ * strings it references.
+ *
+ * The allocator used to initialize the string array is used to deallocate each
+ * string in the array and the array of strings itself.
+ *
+ * \param[inout] array_list object to be finalized
+ * \return `RCUTILS_RET_OK` if successful, or
+ * \return `RCUTILS_RET_INVALID_ARGUMENT` for invalid arguments, or
+ * \return `RCUTILS_RET_ERROR` if an unknown error occurs
+ */
+RCUTILS_PUBLIC
+RCUTILS_WARN_UNUSED
+rcutils_ret_t
+rcutils_array_list_fini(rcutils_array_list_t * array_list);
+
+/// Adds an entry to the list
+/**
+ * This functions adds the provided data to the end of the list. A copy of
+ * the provided data is made to store in the list instead of just storing
+ * the pointer to the provided data.
+ *
+ *
+ * \param[in] array_list to add the data to
+ * \param[in] data a pointer to the data to add to the list
+ * \return `RCUTILS_RET_OK` if successful, or
+ * \return `RCUTILS_RET_INVALID_ARGUMENT` for invalid arguments, or
+ * \return `RCUTILS_RET_BAD_ALLOC` if memory allocation fails, or
+ * \return `RCUTILS_RET_ERROR` if an unknown error occurs
+ */
+RCUTILS_PUBLIC
+RCUTILS_WARN_UNUSED
+rcutils_ret_t
+rcutils_array_list_add(rcutils_array_list_t * array_list, const void * data);
+
+/// Sets an entry in the list to the provided data
+/**
+ * This functions sets the provided data at the specified index in the list. A copy of
+ * the provided data is made to store in the list instead of just storing
+ * the pointer to the provided data.
+ *
+ *
+ * \param[in] array_list to add the data to
+ * \param[in] index the position in the list to set the data
+ * \param[in] data a pointer to the data that will be set in the list
+ * \return `RCUTILS_RET_OK` if successful, or
+ * \return `RCUTILS_RET_INVALID_ARGUMENT` for invalid arguments, or
```

Comment:
@nburek nit: it might be redundant for some, but it'd be nice for the documentation to state that an `index` out of [0, size) bounds also results in `RCUTILS_RET_INVALID_ARGUMENT`.

Reply:
Sure, explicitly added this case to the documentation.